gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,429 @@
+		if err != nil {
+			return err
+		}
+
+		log.Printf("[DEBUG] Project %s delete gitlab project-level approval rule %d", project, ruleIDInt)
+
+		client := meta.(*gitlab.Client)
+		_, err = client.Projects.DeleteProjectApprovalRule(project, ruleIDInt)
+	}
+
+	return err
+}
+
+// getApprovalRuleByID checks the list of rules and finds the one that matches our rule ID.
+func getApprovalRuleByID(meta interface{}, id string) (*gitlab.ProjectApprovalRule, error) {
+	var rules []*gitlab.ProjectApprovalRule
+	var err error
+
+	projectID, ruleIDStr, err := splitApprovalRuleID(id)
+	if err != nil {
+		return nil, err
+	}
+
+	client := meta.(*gitlab.Client)
+	if rules, _, err = client.Projects.GetProjectApprovalRules(projectID); err == nil {
+		var ruleIDInt int
+
+		if ruleIDInt, err = strconv.Atoi(ruleIDStr); err != nil {
+			return nil, err
+		}
+
+		for _, r := range rules {
+			if r.ID == ruleIDInt {
+				log.Printf("[DEBUG] found project-level rule %+v", r)
+				return r, nil
+			}
+		}
+	}
+
+	return nil, err
+}
+
+// createApprovalRuleID creates an ID of two parts: projectID:ruleID
+func createApprovalRuleID(ruleID int, projectID string) string {
+	return buildTwoPartID(gitlab.String(projectID), gitlab.String(strconv.Itoa(ruleID)))
+}
+
+// splitApprovalRuleID splits an approval rule into two parts, projectID:ruleID.
+// An error is returned if one occurs.
+func splitApprovalRuleID(ruleID string) (string, string, error) {
+	return parseTwoPartID(ruleID)
+}
+
+// flattenApprovalRuleUserIDs flattens a list of approval user ids into a list
+// of ints for storage in state.
+func flattenApprovalRuleUserIDs(users []*gitlab.BasicUser) []interface{} {
+	m := []interface{}{}
+	for _, user := range users {
+		m = append(m, user.ID)
+	}
+
+	return m
+}
+
+// flattenApprovalRuleGroupIDs flattens a list of approval group ids into a list
+// of ints for storage in state.
+func flattenApprovalRuleGroupIDs(groups []*gitlab.Group) []interface{} {
+	m := []interface{}{}
+	for _, group := range groups {
+		m = append(m, group.ID)
+	}
+
+	return m
+}
+
+// expandApproverIds Expands an interface into a list of ints to read from state.
+func expandApproverIds(ids interface{}, hasItems bool) []int {
+	m := []int{}
+	if hasItems {
+		for _, id := range ids.(*schema.Set).List() {
+			m = append(m, id.(int))
+		}
+	}
+
+	return m
+}
+
+// getProjectApprovalRuleUserIDs generates a list of approvers for a rule that
+// could not be found in the project or in any attached groups and returns a
+// list containing only the user IDs not found as developer <= higher in the
+// project, or developer <= in any of the rule groups.
+//
+// This method attempts to work around a bug in GitLab where there are instances
+// where the `users` attribute in the response will be empty under certain
+// conditions.
+//
+// Reference: https://gitlab.com/gitlab-org/gitlab/issues/35008
+func getProjectApprovalRuleUserIDs(client *gitlab.Client, ruleID string) ([]int, error) {
```

Comment:
I have to wonder if it's really our responsibility to work around this bug, since doing so adds such complexity and potential for strange behavior, like Matt has pointed out below. How often does it occur? What is the impact to this provider when it does? It seems to me like the impact is that it would show a diff when it shouldn't, but that's not a serious problem because making an update call with the same list of user IDs is safe.